File: go/task.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	args := os.Args
	argc := len(args)
	var primaryArg string
	if argc > 1 {
		primaryArg = args[1]
	} else {
		primaryArg = ""
	}
	switch primaryArg {
	case "ls":
		getRemainingTasks("task.txt", false)
		break
	case "add":
		if argc > 3 {
			setNewTask(args[2], args[3])
		} else {
			fmt.Println("Error: Missing tasks string. Nothing added!")
		}
		break
	case "del":
		if argc > 2 {
			_, err := deleteTask(args[2])
			if err {
				fmt.Println("Error: task with index #" + args[2] + " does not exist. Nothing deleted.")
			} else {
				fmt.Println("Deleted task #" + args[2])
			}
		} else {
			fmt.Println("Error: Missing NUMBER for deleting tasks.")
		}
		break
	case "done":
		if argc > 2 {
			setCompleted(args[2])
		} else {
			fmt.Println("Error: Missing NUMBER for marking tasks as done.")
		}
		break
	case "report":
		getRemainingTasks("task.txt", true)
		fmt.Println()
		getRemainingTasks("completed.txt", true)
		break
	default:
		help()
	}
}

func help() {
	fmt.Println("Usage :-\n$ ./task add 2 hello world    # Add a new item with priority 2 and text \"hello world\" to the list\n$ ./task ls                   # Show incomplete priority list items sorted by priority in ascending order\n$ ./task del INDEX            # Delete the incomplete item with the given index\n$ ./task done INDEX           # Mark the incomplete item with the given index as complete\n$ ./task help                 # Show usage\n$ ./task report               # Statistics")
}

func getFileData(filename string) (data []string) {
	file, err := os.Open(filename)
	data = []string{}
	if err != nil {
		return
	}
	defer file.Close()
	scr := bufio.NewScanner(file)
	for scr.Scan() {
		item := strings.TrimSpace(scr.Text())
		if len(item) == 0 {
			continue
		}
		data = append(data, item)
	}
	return
}

func setFileData(filename string, data []string) {
	file, err := os.Create(filename)
	if err != nil {
		return
	}
	defer file.Close()
	fw := bufio.NewWriter(file)
	for _, line := range data {
		if len(line) == 0 {
			continue
		}
		fmt.Fprintln(fw, line)
	}
	fw.Flush()
}

func getRemainingTasks(filename string, displayQuantity bool) {
	data := getFileData(filename)
	size := len(data)
	if displayQuantity {
		if filename == "completed.txt" {
			fmt.Println("Completed :", size)
		} else {
			fmt.Println("Pending :", size)
		}
	} else if size == 0 {
		fmt.Println("There are no pending tasks!")
	}
	for i := 0; i < size; i++ {
		item := data[i]
		if filename == "completed.txt" {
			fmt.Printf("%d. %s\n", (i + 1), item)
		} else {
			fmt.Printf("%d. %s [%s]\n", (i + 1), item[strings.Index(item, " ")+1:], item[:strings.Index(item, " ")])
		}
	}
}

func setNewTask(priority, task string) {
	data := getFileData("task.txt")
	pr, err := strconv.ParseInt(priority, 0, 32)
	if err != nil {
		fmt.Println("Invalid priority")
		return
	}
	var index int = 0
	for _, item := range data {
		p, err := strconv.ParseInt(item[:strings.Index(item, " ")], 0, 32)
		if err != nil {
			continue
		}
		if pr < p {
			break
		}
		index++
	}
	data = append(data, "")
	copy(data[index+1:], data[index:])
	data[index] = priority + " " + task
	setFileData("task.txt", data)
	fmt.Printf("Added task: \"%s\" with priority %d\n", task, pr)
}

func deleteTask(index string) (deletedItem string, err bool) {
	data := getFileData("task.txt")
	ndx, error := strconv.ParseInt(index, 0, 32)
	if error != nil {
		err = true
		return
	}
	if ndx < 1 || ndx > int64(len(data)) {
		err = true
		return
	}
	deletedItem = data[ndx-1]
	deletedItem = deletedItem[strings.Index(deletedItem, " ")+1:]
	data[ndx-1] = ""
	setFileData("task.txt", data)
	return
}

func setCompleted(index string) {
	deletedItem, error := deleteTask(index)
	if error {
		fmt.Printf("Error: no incomplete item with index #%s exists.\n", index)
		return
	}

	f, err := os.OpenFile("completed.txt", os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	if _, err = f.WriteString(deletedItem + "\n"); err != nil {
		panic(err)
	}
	fmt.Println("Marked item as done.")
}

File: swift/task.swift
import Cocoa

let args = CommandLine.arguments

func help() {
    print("Usage :-\n$ ./task add 2 hello world    # Add a new item with priority 2 and text \"hello world\" to the list\n$ ./task ls                   # Show incomplete priority list items sorted by priority in ascending order\n$ ./task del INDEX            # Delete the incomplete item with the given index\n$ ./task done INDEX           # Mark the incomplete item with the given index as complete\n$ ./task help                 # Show usage\n$ ./task report               # Statistics")
}

let getFileData = { (whichFile: Bool) -> [String] in
  let url = FileManager.default.currentDirectoryPath + (whichFile ? "/completed.txt" : "/task.txt")
  var data: [String] = []
  do {
    let savedData = try String(contentsOfFile: url)
    data = savedData.components(separatedBy: "\n")
    data = data.filter({ !$0.isEmpty })
  } catch { }
  return data
}

let getRemainingTasks = { (whichFile: Bool, displayCount: Bool) in
  let data = getFileData(whichFile)
  let size = data.count
  if displayCount {
    print(whichFile ? "Completed : \(size)" : "Pending : \(size)")
  } else if size == 0 {
    print("There are no pending tasks!")
    return
  }
  for i in 0..<size {
    let item = data[i]
    print(whichFile ? "\(i+1). \(item)" : "\(i+1).\(item[item.firstIndex(of: " ")!...]) [\(item[..<item.firstIndex(of: " ")!])]")
  }
}

let setFileData = { (whichFile: Bool, data: [String]) in
  let lines = data.joined(separator: "\n")
  let url = URL(fileURLWithPath: FileManager.default.currentDirectoryPath + (whichFile ? "/completed.txt" : "/task.txt"))
  do {
    try lines.write(to: url, atomically: false, encoding: .utf8)
  } catch {
    print("Couldn't write to file")
  }
}

let setNewTask = { (priority: String, task: String, dir: String) in
  let pr = Int(priority) ?? -1
  guard pr > -1 else {
    print("Invalid priority")
    return
  }
  var ndx = 0
  var data = getFileData(false)
  data.forEach { item in
    let p = Int(item.prefix(upTo: item.firstIndex(of: " ")!)) ?? -1
    guard p > -1 else {
      return
    }
    if pr<p {
      return
    }
    ndx+=1
  }
  data.insert("\(priority) \(task)", at: ndx)
  setFileData(false, data)
  print("Added task: \"\(task)\" with priority \(priority)")
}

let deleteTask = { (index: String) -> String? in
  let ndx = Int(index) ?? -1
  var data = getFileData(false)
  guard ndx > 0 && ndx <= data.count else {
    return nil
  }
  var deletedItem = data[ndx-1]
  deletedItem = String(deletedItem.suffix(from: deletedItem.firstIndex(of: " ")!)).trimmingCharacters(in: .whitespacesAndNewlines)
  data.remove(at: ndx-1)
  setFileData(false, data)
  return deletedItem
}

let setCompleted = { (index: String) in
  let deletedItem = deleteTask(index)
  guard deletedItem != nil else {
    print("Error: no incomplete item with index #\(index) exists.")
    return
  }
  var data = getFileData(true)
  data.append(deletedItem!)
  setFileData(true, data)
  print("Marked item as done.")
}

let argc = CommandLine.argc
switch argc > 1 ? args[1] : "" {
case "ls":
    getRemainingTasks(false, false)
case "add":
  if (argc > 3) {
    setNewTask(args[2], args[3], args[0])
  } else {
    print("Error: Missing tasks string. Nothing added!")
  }
case "del":
  if (argc > 2) {
    if (deleteTask(args[2]) != nil) {
      print("Deleted task #\(args[2])")
    }
    else {
      print("Error: task with index #\(args[2]) does not exist. Nothing deleted.")
    }
  }
  else {
    print("Error: Missing NUMBER for deleting tasks.")
  }
case "done":
  if (argc > 2) {
    setCompleted(args[2])
  }
  else {
    print("Error: Missing NUMBER for marking tasks as done.")
  }
case "report":
  getRemainingTasks(false, true)
  print()
  getRemainingTasks(true, true)
default:
    help()
}